File: packages/validator-ts/src/pattern/recognizer.ts
/**
 * Recognizer of the CRUDUI pattern language (validation-rules.md, "Patterns" and
 * "Parameter errors").
 *
 * It reads a pattern as code points, left to right, and either returns its
 * syntax tree or throws a `PatternSyntaxError` for the first invalid construct,
 * with the reason and code-point offset the specification assigns to it.
 */

import type { Alternation, Atom, ClassMember, CodePoint, Property, Quantifier, Shorthand, Term } from './ast';
import { PatternSyntaxError } from './errors';
import { isGeneralCategory, isScript } from './sets';

/** Largest quantifier bound. */
export const QUANTIFIER_LIMIT = 1000;

/** Largest pattern size. */
export const SIZE_LIMIT = 1000;

/** Deepest group nesting. */
export const DEPTH_LIMIT = 100;

/** A recognized pattern. Leading `^` and trailing `$` add nothing and are not kept. */
export interface RecognizedPattern {
  body: Alternation;
}

const cp = (char: string): number => char.codePointAt(0)!;

/** Characters a `\` escape turns into themselves. */
const SYNTAX_ESCAPES = new Set([...'^$\\.*+?()[]{}|/-'].map(cp));

/** Control escapes. */
const CONTROL_ESCAPES = new Map<number, number>([
  [cp('t'), 0x09],
  [cp('n'), 0x0a],
  [cp('r'), 0x0d],
  [cp('f'), 0x0c],
  [cp('v'), 0x0b],
]);

const SHORTHANDS = new Map<number, { set: Shorthand; negated: boolean }>([
  [cp('d'), { set: 'digit', negated: false }],
  [cp('w'), { set: 'word', negated: false }],
  [cp('s'), { set: 'space', negated: false }],
  [cp('D'), { set: 'digit', negated: true }],
  [cp('W'), { set: 'word', negated: true }],
  [cp('S'), { set: 'space', negated: true }],
]);

/** Characters that are never literals outside a bracket class. */
const NOT_LITERAL = new Set([...'\\^$.|?*+()[]{}'].map(cp));
const QUANTIFIER_START = new Set([...'*+?{'].map(cp));

const BACKSLASH = cp('\\');
const OPEN_GROUP = cp('(');
const CLOSE_GROUP = cp(')');
const OPEN_CLASS = cp('[');
const CLOSE_CLASS = cp(']');
const OPEN_BRACE = cp('{');
const CLOSE_BRACE = cp('}');
const CARET = cp('^');
const DOLLAR = cp('$');
const DOT = cp('.');
const BAR = cp('|');
const QUESTION = cp('?');
const STAR = cp('*');
const PLUS = cp('+');
const DASH = cp('-');
const COMMA = cp(',');
const COLON = cp(':');
const LESS = cp('<');
const GREATER = cp('>');
const EQUALS = cp('=');
const BANG = cp('!');

const isSurrogate = (c: number): boolean => c >= 0xd800 && c <= 0xdfff;
const isDigit = (c: number | undefined): c is number => c !== undefined && c >= 0x30 && c <= 0x39;
const isHex = (c: number | undefined): c is number =>
  c !== undefined && (isDigit(c) || (c >= 0x41 && c <= 0x46) || (c >= 0x61 && c <= 0x66));
const isNameStart = (c: number | undefined): boolean =>
  c !== undefined && ((c >= 0x41 && c <= 0x5a) || (c >= 0x61 && c <= 0x7a) || c === 0x5f);
const isNamePart = (c: number | undefined): boolean => isNameStart(c) || isDigit(c);

/** An escape read by the recognizer: one code point, a shorthand set or a property. */
type Escape = CodePoint | { kind: 'shorthand'; set: Shorthand; negated: boolean } | Property;

class Recognizer {
  private readonly source: number[];
  private position = 0;
  private readonly groupNames = new Set<string>();
  private depth = 0;

  constructor(pattern: string) {
    this.source = Array.from(pattern, cp);
  }

  recognize(): RecognizedPattern {
    const length = this.source.length;
    if (length === 0) throw new PatternSyntaxError('empty pattern', 0);
    if (this.source[0] === CARET) this.position = 1;
    const alternation = this.alternation();
    if (this.position < length) {
      // Only a `)` without an open group stops the top-level alternation early.
      throw new PatternSyntaxError('unexpected character', this.position);
    }
    // The size limit is checked once the rest of the pattern is valid.
    if (alternationSize(alternation) > SIZE_LIMIT) throw new PatternSyntaxError('pattern too large', 0);
    return { body: alternation };
  }

  private peek(offset = 0): number | undefined {
    return this.source[this.position + offset];
  }

  /** Alternatives up to `)` or the end. */
  private alternation(): Alternation {
    const branches: Term[][] = [this.sequence()];
    while (this.peek() === BAR) {
      this.position++;
      branches.push(this.sequence());
    }
    return { branches };
  }

  private sequence(): Term[] {
    const terms: Term[] = [];
    const length = this.source.length;
    while (this.position < length) {
      const char = this.source[this.position];
      if (char === BAR || char === CLOSE_GROUP) break;
      if (char === DOLLAR || char === CARET) {
        // `^` at 0 was consumed before; `$` is an anchor only as the last character.
        if (char === CARET || this.position !== length - 1) {
          throw new PatternSyntaxError('unexpected character', this.position);
        }
        this.position++;
        break;
      }
      if (QUANTIFIER_START.has(char)) throw new PatternSyntaxError('invalid quantifier', this.position);
      const atom = this.atom();
      terms.push({ atom, quantifier: this.quantifier() });
    }
    return terms;
  }

  private atom(): Atom {
    const start = this.position;
    const char = this.source[start];
    switch (char) {
      case OPEN_GROUP:
        return this.group();
      case OPEN_CLASS:
        return this.bracketClass();
      case DOT:
        this.position++;
        return { kind: 'any' };
      case BACKSLASH:
        return this.escape(null);
    }
    if (NOT_LITERAL.has(char) || isSurrogate(char)) throw new PatternSyntaxError('unexpected character', start);
    this.position++;
    return { kind: 'char', value: char };
  }

  private group(): Atom {
    const start = this.position;
    if (this.depth === DEPTH_LIMIT) throw new PatternSyntaxError('nesting too deep', start);
    this.position++;
    if (this.peek() === QUESTION) {
      const next = this.peek(1);
      if (next === COLON) {
        this.position += 2;
      } else if (next === LESS && this.peek(2) !== EQUALS && this.peek(2) !== BANG) {
        this.position += 2;
        this.groupName(start);
      } else {
        throw new PatternSyntaxError('unsupported construct', start);
      }
    }
    this.depth++;
    const body = this.alternation();
    this.depth--;
    if (this.position >= this.source.length) {
      throw new PatternSyntaxError('unterminated group', this.source.length);
    }
    this.position++; // `)`
    return { kind: 'group', body };
  }

  /** `name>` after `(?<`. */
  private groupName(groupStart: number): void {
    const nameStart = this.position;
    if (!isNameStart(this.peek())) throw new PatternSyntaxError('invalid group name', groupStart);
    while (isNamePart(this.peek())) this.position++;
    if (this.peek() !== GREATER) throw new PatternSyntaxError('invalid group name', groupStart);
    const name = String.fromCodePoint(...this.source.slice(nameStart, this.position));
    this.position++;
    if (this.groupNames.has(name)) throw new PatternSyntaxError('duplicate group name', groupStart);
    this.groupNames.add(name);
  }

  /** An optional quantifier with its optional lazy `?`. */
  private quantifier(): Quantifier | null {
    const start = this.position;
    const char = this.peek();
    let min: number;
    let max: number | null;
    if (char === STAR) {
      [min, max] = [0, null];
      this.position++;
    } else if (char === PLUS) {
      [min, max] = [1, null];
      this.position++;
    } else if (char === QUESTION) {
      [min, max] = [0, 1];
      this.position++;
    } else if (char === OPEN_BRACE) {
      this.position++;
      min = this.bound(start);
      if (this.peek() === CLOSE_BRACE) {
        max = min;
      } else if (this.peek() === COMMA) {
        this.position++;
        max = this.peek() === CLOSE_BRACE ? null : this.bound(start);
        if (max !== null && min > max) throw new PatternSyntaxError('invalid quantifier', start);
      } else {
        throw new PatternSyntaxError('invalid quantifier', start);
      }
      if (this.peek() !== CLOSE_BRACE) throw new PatternSyntaxError('invalid quantifier', start);
      this.position++;
    } else {
      return null;
    }
    const lazy = this.peek() === QUESTION;
    if (lazy) this.position++;
    return { min, max, lazy };
  }

  /** A decimal bound of at most `QUANTIFIER_LIMIT`. */
  private bound(quantifierStart: number): number {
    const digitsStart = this.position;
    let value = 0;
    while (isDigit(this.peek())) {
      value = Math.min(value * 10 + (this.peek()! - 0x30), QUANTIFIER_LIMIT + 1);
      this.position++;
    }
    if (this.position === digitsStart || value > QUANTIFIER_LIMIT) {
      throw new PatternSyntaxError('invalid quantifier', quantifierStart);
    }
    return value;
  }

  /**
   * An escape at the current `\`. Inside a bracket class (`classStart` is the
   * offset of its `[`), `\D`, `\W` and `\S` are an invalid class.
   */
  private escape(classStart: number | null): Escape {
    const start = this.position;
    const kind = this.peek(1);
    if (kind === undefined) throw new PatternSyntaxError('invalid escape', start);
    if (SYNTAX_ESCAPES.has(kind)) {
      this.position += 2;
      return { kind: 'char', value: kind };
    }
    const control = CONTROL_ESCAPES.get(kind);
    if (control !== undefined) {
      this.position += 2;
      return { kind: 'char', value: control };
    }
    const shorthand = SHORTHANDS.get(kind);
    if (shorthand) {
      if (shorthand.negated && classStart !== null) throw new PatternSyntaxError('invalid class', classStart);
      this.position += 2;
      return { kind: 'shorthand', ...shorthand };
    }
    if (kind === cp('x')) {
      if (!isHex(this.peek(2)) || !isHex(this.peek(3))) throw new PatternSyntaxError('invalid escape', start);
      const value = parseInt(String.fromCodePoint(this.peek(2)!, this.peek(3)!), 16);
      this.position += 4;
      return { kind: 'char', value };
    }
    if (kind === cp('u')) return this.codePointEscape(start);
    if (kind === cp('p') || kind === cp('P')) return this.property(start, kind === cp('P'));
    throw new PatternSyntaxError('invalid escape', start);
  }

  /** `\u{H…}`: 1–6 hexadecimal digits naming a Unicode scalar value. */
  private codePointEscape(start: number): CodePoint {
    if (this.peek(2) !== OPEN_BRACE) throw new PatternSyntaxError('invalid escape', start);
    let index = 3;
    let digits = '';
    while (isHex(this.peek(index))) {
      digits += String.fromCodePoint(this.peek(index)!);
      index++;
    }
    if (digits.length < 1 || digits.length > 6 || this.peek(index) !== CLOSE_BRACE) {
      throw new PatternSyntaxError('invalid escape', start);
    }
    const value = parseInt(digits, 16);
    if (value > 0x10ffff || isSurrogate(value)) {
      throw new PatternSyntaxError('invalid escape', start);
    }
    this.position += index + 1;
    return { kind: 'char', value };
  }

  /** `\p{X}` or `\P{X}` with a general category or `Script=Name`. */
  private property(start: number, negated: boolean): Property {
    if (this.peek(2) !== OPEN_BRACE) throw new PatternSyntaxError('invalid property', start);
    let index = 3;
    while (this.peek(index) !== undefined && this.peek(index) !== CLOSE_BRACE) index++;
    if (this.peek(index) === undefined) throw new PatternSyntaxError('invalid property', start);
    const text = String.fromCodePoint(...this.source.slice(this.position + 3, this.position + index));
    let property: Property;
    if (isGeneralCategory(text)) {
      property = { kind: 'property', negated, name: text, script: false };
    } else if (text.startsWith('Script=') && isScript(text.slice('Script='.length))) {
      property = { kind: 'property', negated, name: text.slice('Script='.length), script: true };
    } else {
      throw new PatternSyntaxError('invalid property', start);
    }
    this.position += index + 1;
    return property;
  }

  /** `[…]` or `[^…]` with at least one member. */
  private bracketClass(): Atom {
    const start = this.position;
    const length = this.source.length;
    this.position++;
    const negated = this.peek() === CARET;
    if (negated) this.position++;
    const firstMember = this.position;
    const members: ClassMember[] = [];
    for (;;) {
      if (this.position >= length) throw new PatternSyntaxError('unterminated class', length);
      if (this.peek() === CLOSE_CLASS) {
        this.position++;
        break;
      }
      const memberStart = this.position;
      const from = this.classAtom(start, memberStart === firstMember);
      if (this.peek() === DASH && this.peek(1) !== undefined && this.peek(1) !== CLOSE_CLASS) {
        // A range `from-to` of single code points in non-descending order. The
        // class is read left to right: the end is read before the range is judged.
        this.position++;
        const to = this.classAtom(start, false);
        if (from.kind !== 'char' || to.kind !== 'char' || to.value < from.value) {
          throw new PatternSyntaxError('invalid range', memberStart);
        }
        members.push({ kind: 'range', from: from.value, to: to.value });
      } else {
        members.push(from.kind === 'shorthand' ? { kind: 'shorthand', set: from.set } : from);
      }
    }
    if (members.length === 0) throw new PatternSyntaxError('invalid class', start);
    return { kind: 'class', negated, members };
  }

  /**
   * One class member: a literal, an escape, a shorthand or a property. `[` and
   * `\` are never literals; `-` is a literal only at a boundary of the class.
   */
  private classAtom(classStart: number, first: boolean): Escape {
    const char = this.peek()!;
    if (char === OPEN_CLASS) throw new PatternSyntaxError('invalid class', classStart);
    if (char === BACKSLASH) return this.escape(classStart);
    if (char === DASH) {
      const next = this.peek(1);
      const boundary = first || next === CLOSE_CLASS || next === undefined;
      if (!boundary) throw new PatternSyntaxError('invalid class', classStart);
    }
    if (isSurrogate(char)) throw new PatternSyntaxError('unexpected character', this.position);
    this.position++;
    return { kind: 'char', value: char };
  }
}

/**
 * Recognize a pattern of the CRUDUI pattern language.
 *
 * @throws {PatternSyntaxError} for the first construct outside the language.
 */
export function recognizePattern(pattern: string): RecognizedPattern {
  return new Recognizer(pattern).recognize();
}

/** Sizes above the limit are kept at limit + 1, so products never overflow. */
const cap = (size: number): number => Math.min(size, SIZE_LIMIT + 1);

function alternationSize(node: Alternation): number {
  let size = 0;
  for (const terms of node.branches) {
    for (const term of terms) size = cap(size + termSize(term));
  }
  return size;
}

/**
 * The size of a term: an atom is 1, a group its body; a quantified item is
 * multiplied by its maximum, or by its minimum plus one when unbounded.
 */
export function termSize(term: Term): number {
  const item = term.atom.kind === 'group' ? alternationSize(term.atom.body) : 1;
  const quantifier = term.quantifier;
  if (!quantifier) return item;
  return cap(item * (quantifier.max === null ? quantifier.min + 1 : quantifier.max));
}

/** The size of a sequence or alternation body. */
export { alternationSize };

File: packages/validator-rust/src/validate/pattern/recognizer.rs
//! Recognizer of the CRUDUI pattern language (validation rules, "Patterns" and
//! "Parameter errors").
//!
//! It reads the pattern left to right and reports the first construct outside the
//! language with its reason and code-point offset. The pattern size is checked
//! later, once the whole pattern is otherwise valid.

use std::collections::HashSet;

use crate::validate::unicode::{lookup, GENERAL_CATEGORIES, SCRIPTS};

use super::syntax::{Atom, ClassItem, PatternError, Property, Reason, Shorthand, Token};

/// The largest quantifier bound.
const BOUND_LIMIT: u64 = 1000;

/// The deepest group nesting.
const DEPTH_LIMIT: usize = 100;

/// Characters that `\` makes literal.
const ESCAPABLE: &[char] = &[
    '^', '$', '\\', '.', '*', '+', '?', '(', ')', '[', ']', '{', '}', '|', '/', '-',
];

/// Recognize `source`, returning its tokens or the first construct outside the language.
pub(crate) fn recognize(source: &str) -> Result<Vec<Token>, PatternError> {
    Recognizer {
        chars: source.chars().collect(),
        pos: 0,
        tokens: Vec::new(),
        names: HashSet::new(),
    }
    .run()
}

struct Recognizer {
    chars: Vec<char>,
    pos: usize,
    tokens: Vec<Token>,
    names: HashSet<String>,
}

fn error(reason: Reason, offset: usize) -> PatternError {
    PatternError { reason, offset }
}

fn is_quantifier_start(c: char) -> bool {
    matches!(c, '*' | '+' | '?' | '{')
}

impl Recognizer {
    fn peek(&self) -> Option<char> {
        self.chars.get(self.pos).copied()
    }

    fn peek_at(&self, offset: usize) -> Option<char> {
        self.chars.get(self.pos + offset).copied()
    }

    fn run(mut self) -> Result<Vec<Token>, PatternError> {
        let len = self.chars.len();
        if len == 0 {
            return Err(error(Reason::EmptyPattern, 0));
        }
        // The number of open groups.
        let mut depth = 0usize;
        while let Some(c) = self.peek() {
            let start = self.pos;
            // Whether the construct just read is an atom or group a quantifier may follow.
            let quantifiable = match c {
                '^' if start == 0 => {
                    self.pos += 1;
                    false
                }
                '$' if start == len - 1 => {
                    self.pos += 1;
                    false
                }
                '^' | '$' | ']' | '}' => return Err(error(Reason::UnexpectedCharacter, start)),
                '|' => {
                    self.pos += 1;
                    self.tokens.push(Token::Alternate);
                    false
                }
                '(' => {
                    if depth == DEPTH_LIMIT {
                        return Err(error(Reason::NestingTooDeep, start));
                    }
                    self.open_group()?;
                    depth += 1;
                    false
                }
                ')' => {
                    if depth == 0 {
                        return Err(error(Reason::UnexpectedCharacter, start));
                    }
                    self.pos += 1;
                    self.tokens.push(Token::Close);
                    depth -= 1;
                    true
                }
                c if is_quantifier_start(c) => {
                    return Err(error(Reason::InvalidQuantifier, start));
                }
                '[' => {
                    let atom = self.class()?;
                    self.atom(atom)
                }
                '.' => {
                    self.pos += 1;
                    self.atom(Atom::Any)
                }
                '\\' => {
                    let atom = self.escape()?;
                    self.atom(atom)
                }
                c => {
                    self.pos += 1;
                    self.atom(Atom::Literal(c))
                }
            };
            if quantifiable {
                self.quantifier()?;
            }
        }
        if depth > 0 {
            return Err(error(Reason::UnterminatedGroup, len));
        }
        Ok(self.tokens)
    }

    fn atom(&mut self, atom: Atom) -> bool {
        self.tokens.push(Token::Atom(atom));
        true
    }

    /// An optional quantifier, with an optional lazy `?` that changes nothing for a
    /// whole match.
    fn quantifier(&mut self) -> Result<(), PatternError> {
        let (min, max) = match self.peek() {
            Some('{') => self.braces()?,
            Some(c @ ('*' | '+' | '?')) => {
                self.pos += 1;
                match c {
                    '*' => (0, None),
                    '+' => (1, None),
                    _ => (0, Some(1)),
                }
            }
            _ => return Ok(()),
        };
        if self.peek() == Some('?') {
            self.pos += 1;
        }
        if let Some(c) = self.peek() {
            if is_quantifier_start(c) {
                return Err(error(Reason::InvalidQuantifier, self.pos));
            }
        }
        self.tokens.push(Token::Quantifier { min, max });
        Ok(())
    }

    /// `{n}`, `{n,}` or `{n,m}` with n ≤ m ≤ 1000.
    fn braces(&mut self) -> Result<(u32, Option<u32>), PatternError> {
        let start = self.pos;
        let invalid = error(Reason::InvalidQuantifier, start);
        self.pos += 1;
        let min = self.number().ok_or(invalid)?;
        let max = match self.peek() {
            Some('}') => Some(min),
            Some(',') => {
                self.pos += 1;
                if self.peek() == Some('}') {
                    None
                } else {
                    Some(self.number().ok_or(invalid)?)
                }
            }
            _ => return Err(invalid),
        };
        if self.peek() != Some('}') {
            return Err(invalid);
        }
        self.pos += 1;
        if min > BOUND_LIMIT || max.is_some_and(|max| max > BOUND_LIMIT || max < min) {
            return Err(invalid);
        }
        // Both bounds are at most 1000 here.
        Ok((min as u32, max.map(|max| max as u32)))
    }

    /// One or more decimal digits.
    fn number(&mut self) -> Option<u64> {
        let start = self.pos;
        let mut value: u64 = 0;
        while let Some(digit) = self.peek().and_then(|c| c.to_digit(10)) {
            value = value.saturating_mul(10).saturating_add(u64::from(digit));
            self.pos += 1;
        }
        (self.pos > start).then_some(value)
    }

    /// `(…)`, `(?:…)` or `(?<name>…)`; pushes the open token.
    fn open_group(&mut self) -> Result<(), PatternError> {
        let start = self.pos;
        self.pos += 1;
        if self.peek() == Some('?') {
            match (self.peek_at(1), self.peek_at(2)) {
                (Some(':'), _) => self.pos += 2,
                (Some('<'), next) if !matches!(next, Some('=') | Some('!')) => {
                    self.pos += 2;
                    let name_start = self.pos;
                    while self.peek().is_some_and(|c| c != '>') {
                        self.pos += 1;
                    }
                    if self.peek().is_none() {
                        return Err(error(Reason::InvalidGroupName, start));
                    }
                    let name: String = self.chars[name_start..self.pos].iter().collect();
                    self.pos += 1;
                    if !is_group_name(&name) {
                        return Err(error(Reason::InvalidGroupName, start));
                    }
                    if !self.names.insert(name) {
                        return Err(error(Reason::DuplicateGroupName, start));
                    }
                }
                _ => return Err(error(Reason::UnsupportedConstruct, start)),
            }
        }
        self.tokens.push(Token::Open);
        Ok(())
    }

    /// An escape outside a bracket class.
    fn escape(&mut self) -> Result<Atom, PatternError> {
        match self.peek_at(1) {
            Some(c @ ('d' | 'w' | 's' | 'D' | 'W' | 'S')) => {
                self.pos += 2;
                Ok(Atom::Shorthand {
                    set: shorthand(c),
                    negated: c.is_ascii_uppercase(),
                })
            }
            Some('p' | 'P') => self.property().map(Atom::Property),
            _ => self.literal_escape().map(Atom::Literal),
        }
    }

    /// A literal, control, `\xHH` or `\u{H…}` escape at the current position.
    fn literal_escape(&mut self) -> Result<char, PatternError> {
        let start = self.pos;
        let invalid = error(Reason::InvalidEscape, start);
        let c = self.peek_at(1).ok_or(invalid)?;
        self.pos += 2;
        match c {
            c if ESCAPABLE.contains(&c) => Ok(c),
            't' => Ok('\t'),
            'n' => Ok('\n'),
            'r' => Ok('\r'),
            'f' => Ok('\u{C}'),
            'v' => Ok('\u{B}'),
            'x' => {
                let high = self.peek().and_then(|c| c.to_digit(16)).ok_or(invalid)?;
                let low = self
                    .peek_at(1)
                    .and_then(|c| c.to_digit(16))
                    .ok_or(invalid)?;
                self.pos += 2;
                char::from_u32(high * 16 + low).ok_or(invalid)
            }
            'u' => {
                if self.peek() != Some('{') {
                    return Err(invalid);
                }
                self.pos += 1;
                let mut value: u32 = 0;
                let mut count = 0;
                while let Some(digit) = self.peek().and_then(|c| c.to_digit(16)) {
                    count += 1;
                    if count > 6 {
                        return Err(invalid);
                    }
                    value = value * 16 + digit;
                    self.pos += 1;
                }
                if count == 0 || self.peek() != Some('}') {
                    return Err(invalid);
                }
                self.pos += 1;
                char::from_u32(value).ok_or(invalid)
            }
            _ => Err(invalid),
        }
    }

    /// `\p{X}`, `\P{X}`, `\p{Script=Name}` or `\P{Script=Name}` at the current position.
    fn property(&mut self) -> Result<Property, PatternError> {
        let start = self.pos;
        let invalid = error(Reason::InvalidProperty, start);
        let negated = self.peek_at(1) == Some('P');
        if self.peek_at(2) != Some('{') {
            return Err(invalid);
        }
        self.pos += 3;
        let name_start = self.pos;
        while self.peek().is_some_and(|c| c != '}') {
            self.pos += 1;
        }
        if self.peek().is_none() {
            return Err(invalid);
        }
        let name: String = self.chars[name_start..self.pos].iter().collect();
        self.pos += 1;
        let ranges = match name.strip_prefix("Script=") {
            Some(script) => lookup(SCRIPTS, script),
            None => lookup(GENERAL_CATEGORIES, &name),
        }
        .ok_or(invalid)?;
        Ok(Property { negated, ranges })
    }

    /// `[…]` or `[^…]`.
    fn class(&mut self) -> Result<Atom, PatternError> {
        let start = self.pos;
        let len = self.chars.len();
        self.pos += 1;
        let negated = self.peek() == Some('^');
        if negated {
            self.pos += 1;
        }
        let mut items: Vec<ClassItem> = Vec::new();
        loop {
            let item_start = self.pos;
            let item = match self.peek() {
                None => return Err(error(Reason::UnterminatedClass, len)),
                Some(']') if items.is_empty() => return Err(error(Reason::InvalidClass, start)),
                Some(']') => {
                    self.pos += 1;
                    return Ok(Atom::Class { negated, items });
                }
                // A boundary `-` is a literal: first, or last before `]`.
                Some('-') if items.is_empty() || self.peek_at(1) == Some(']') => {
                    self.pos += 1;
                    ClassItem::Char('-')
                }
                _ => self.class_member(start)?,
            };
            // A non-boundary `-` after a member makes a range.
            if self.peek() == Some('-') && !matches!(self.peek_at(1), Some(']') | None) {
                self.pos += 1;
                let end = self.class_member(start)?;
                match (item, end) {
                    (ClassItem::Char(low), ClassItem::Char(high)) if low <= high => {
                        items.push(ClassItem::Range(low, high));
                    }
                    _ => return Err(error(Reason::InvalidRange, item_start)),
                }
                // A `-` after a range is neither at a boundary nor a range operator.
                if self.peek() == Some('-') && !matches!(self.peek_at(1), Some(']') | None) {
                    return Err(error(Reason::InvalidClass, start));
                }
            } else {
                items.push(item);
            }
        }
    }

    /// One class member or range endpoint; `class_start` locates class errors.
    fn class_member(&mut self, class_start: usize) -> Result<ClassItem, PatternError> {
        match self.peek() {
            Some('[') => Err(error(Reason::InvalidClass, class_start)),
            Some('\\') => match self.peek_at(1) {
                Some('D' | 'W' | 'S') => Err(error(Reason::InvalidClass, class_start)),
                Some(c @ ('d' | 'w' | 's')) => {
                    self.pos += 2;
                    Ok(ClassItem::Shorthand(shorthand(c)))
                }
                Some('p' | 'P') => self.property().map(ClassItem::Property),
                _ => self.literal_escape().map(ClassItem::Char),
            },
            Some(c) => {
                self.pos += 1;
                Ok(ClassItem::Char(c))
            }
            None => Err(error(Reason::UnterminatedClass, self.chars.len())),
        }
    }
}

fn shorthand(c: char) -> Shorthand {
    match c.to_ascii_lowercase() {
        'd' => Shorthand::Digit,
        'w' => Shorthand::Word,
        _ => Shorthand::Space,
    }
}

/// `[A-Za-z_][A-Za-z0-9_]*`
fn is_group_name(name: &str) -> bool {
    let mut chars = name.chars();
    chars
        .next()
        .is_some_and(|c| c.is_ascii_alphabetic() || c == '_')
        && chars.all(|c| c.is_ascii_alphanumeric() || c == '_')
}
